store/tikv/client.go

```diff
@@ -48,102 +58,265 @@ const (
+		var err error
+		resp, err = client.KvPrewrite(ctx, req)
+		if err != nil {
+			conn.Close()
+		}
+		return err
+	}
+	if err := c.callFunc(addr, f, "kv"); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return resp, nil
+}
 
-	select {
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	default:
+func (c *rpcClient) KvCommit(ctx goctx.Context, addr string, req *kvrpcpb.CommitRequest) (*kvrpcpb.CommitResponse, error) {
+	var resp *kvrpcpb.CommitResponse
+	f := func(conn *Conn) error {
+		client := tikvpb.NewTiKVClient(conn.ClientConn)
+		var err error
+		resp, err = client.KvCommit(ctx, req)
+		if err != nil {
+			conn.Close()
+		}
+		return err
+	}
+	if err := c.callFunc(addr, f, "kv"); err != nil {
+		return nil, errors.Trace(err)
 	}
+	return resp, nil
+}
 
-	conn, err := c.p.GetConn(addr)
-	if err != nil {
+func (c *rpcClient) KvCleanup(ctx goctx.Context, addr string, req *kvrpcpb.CleanupRequest) (*kvrpcpb.CleanupResponse, error) {
+	var resp *kvrpcpb.CleanupResponse
+	f := func(conn *Conn) error {
+		client := tikvpb.NewTiKVClient(conn.ClientConn)
+		var err error
+		resp, err = client.KvCleanup(ctx, req)
+		if err != nil {
+			conn.Close()
+		}
+		return err
+	}
+	if err := c.callFunc(addr, f, "kv"); err != nil {
 		return nil, errors.Trace(err)
 	}
-	defer c.p.PutConn(conn)
-	msg := msgpb.Message{
-		MsgType: msgpb.MessageType_KvReq,
-		KvReq:   req,
+	return resp, nil
+}
+
+func (c *rpcClient) KvBatchGet(ctx goctx.Context, addr string, req *kvrpcpb.BatchGetRequest) (*kvrpcpb.BatchGetResponse, error) {
+	var resp *kvrpcpb.BatchGetResponse
+	f := func(conn *Conn) error {
+		client := tikvpb.NewTiKVClient(conn.ClientConn)
+		var err error
+		resp, err = client.KvBatchGet(ctx, req)
+		if err != nil {
+			conn.Close()
+		}
+		return err
 	}
-	err = c.doSend(conn, &msg, writeTimeout, timeout)
-	if err != nil {
-		conn.Close()
+	if err := c.callFunc(addr, f, "kv"); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return resp, nil
+}
+
+func (c *rpcClient) KvBatchRollback(ctx goctx.Context, addr string, req *kvrpcpb.BatchRollbackRequest) (*kvrpcpb.BatchRollbackResponse, error) {
+	var resp *kvrpcpb.BatchRollbackResponse
+	f := func(conn *Conn) error {
+		client := tikvpb.NewTiKVClient(conn.ClientConn)
+		var err error
+		resp, err = client.KvBatchRollback(ctx, req)
+		if err != nil {
+			conn.Close()
+		}
+		return err
+	}
+	if err := c.callFunc(addr, f, "kv"); err != nil {
 		return nil, errors.Trace(err)
 	}
-	if msg.GetMsgType() != msgpb.MessageType_KvResp || msg.GetKvResp() == nil {
-		conn.Close()
-		return nil, errors.Trace(errInvalidResponse)
+	return resp, nil
+}
+
+func (c *rpcClient) KvScanLock(ctx goctx.Context, addr string, req *kvrpcpb.ScanLockRequest) (*kvrpcpb.ScanLockResponse, error) {
+	var resp *kvrpcpb.ScanLockResponse
+	f := func(conn *Conn) error {
+		client := tikvpb.NewTiKVClient(conn.ClientConn)
+		var err error
+		resp, err = client.KvScanLock(ctx, req)
+		if err != nil {
+			conn.Close()
```

Comment:
I prefer to close connection when error happens in function `callFunc`, since the connection would been closed in all `f func(conn *Conn) error` with the same conditions meet.